### ISSUE TITLE
Loosing connection displays unhelpful internal error message

### ISSUE THREAD

**Issue by Dravere** [NodeBB/NodeBB#6410]
- **NodeBB version:** 1.8.1
- **NodeBB git hash:** ad8e121
- **Database type:** mongo
- **Database version:** 3.6.2
- **Exact steps to cause this issue:**
    1. Visit your forum
    2. Disable your internet connection
    3. Click on any link of your forum
- **What you expected:**
    * Either a never ending loading element
    * Or an information that I have no connection
- **What happened instead:** I got an "Internal Error". Which makes it look like there was an error on the server. Not really helpful.

I had a quick glance at the code. Could it be because of this line?
https://github.com/NodeBB/NodeBB/blob/ad8e121/public/src/ajaxify.js#L320

Basically if the Ajax request fails because of a connection issue it will rewrite it as an HTTP error 500.

**Comment by julianlam** [comment]
Hi @Dravere -- unfortunately I cannot reproduce this bug. When I browse to a page, and then open the dev console, and check "offline" in the network tab, subsequent page requests show the loading bar and a blank page, but not an "internal error".

**Comment by barisusakli** [comment]
I can repro on http://community.nodebb.org by going offline and clicking /recent.

![image](https://user-images.githubusercontent.com/2583394/38201283-bd4a9096-3665-11e8-982e-13dbb9ae2840.png)

**Comment by Dravere** [comment]
@julianlam Thanks for mentioning that offline checkbox. So far I always disabled my network interface to test offline stuff. This is a neat feature :smile: 

But like @barisusakli mentioned I can reproduce it on https://community.nodebb.org/:
1. Go to the forum
2. Go offline
3. Click on any of the categories, any of the latest posts, your profile or recent, etc. and I get presented with this error message.

I can produce this error with the latest Vivaldi [1.14.1077.60] (thus Chromium [64.0.3282.189]) and latest Firefox [59.0.2].

**Comment by julianlam** [comment]
Yes... as can I -- I wonder what the problem is, as it is not exhibiting on our local dev environments.

**Comment by Dravere** [comment]
So I did some tests as well and managed to get an instance that is not exhibiting this problem as well. I tried to follow the problem to its roots but I have mostly no clue at the moment. The only difference I could find is this error popping up on the instance that doesn't have this error:
```
require.js:7 Uncaught Error: Script error for "/forum/assets/templates/500.js"
http://requirejs.org/docs/errors.html#scripterror
    at F (require.js:7)
    at HTMLScriptElement.onScriptError (require.js:30)
F @ require.js:7
onScriptError @ require.js:30
setTimeout (async)
p @ require.js:13
(anonymous) @ require.js:26
setTimeout (async)
g.nextTick @ require.js:32
c @ require.js:26
requirejs @ require.js:32
ajaxify.loadTemplate @ ajaxify.js:332
(anonymous) @ benchpress.js:163
a @ benchpress.js:162
(anonymous) @ benchpress.js:184
(anonymous) @ benchpress.js:77
e @ benchpress.js:76
s @ benchpress.js:183
c @ benchpress.js:222
l @ ajaxify.js:185
c @ ajaxify.js:164
(anonymous) @ ajaxify.js:92
error @ ajaxify.js:323
c @ jquery.js:3268
fireWith @ jquery.js:3398
C @ jquery.js:9307
(anonymous) @ jquery.js:9540
XMLHttpRequest.send (async)
send @ jquery.js:9600
ajax @ jquery.js:9206
ajaxify.loadData @ ajaxify.js:291
ajaxify.go @ ajaxify.js:86
s @ ajaxify.js:362
(anonymous) @ ajaxify.js:413
dispatch @ jquery.js:5183
g.handle @ jquery.js:4991
```
(this instance is located in a sub-directory domain.tld/forum/...)

But honestly I can't make a lot from that. I also found out that the ones displaying this error apparently load the 500 template from the page itself located at the bottom:
```
window.addEventListener('load', function () {
		define('/assets/templates/500.js', function () {
			function compiled(helpers, context, get, iter, helper) {
				return '<div class="alert alert-danger">\n\t<strong>Internal Error.</strong>\n\t<p>Oops! Looks like something went wrong!</p>\n\t<p>' + 
					helpers.__escape(get(context && context['path'])) + '</p>\n\t' + 
					(get(context && context['error']) ? '<p>' + helpers.__escape(get(context && context['error'])) + '</p>' : '') + '\n\n\t' + 
					(get(context && context['returnLink']) ? '\n\t<p>Press back to return to the previous page</p>\n\t' : '') + '\n</div>\n';
			}

			return compiled;
		});
	});
```

The ones not displaying this error never use that definition. They can't find that definition and try to load the online one which then never finishes with the promises in the Benchpress and thus looks correct? But I'm confused as to why they can't find that definition. I see no difference.

But all instances try to load the `/assets/template/500.js` template in case of no connection. Which would be wrong anyway. For some reasons some instances try to load it from the web while others use the one provided in the page. I tried to disable the cache, hard-refresh the pages, etc. but the difference remains the same.

**Comment by nodebb-misty** [comment]
Issue closed due to inactivity.

----

*This is an automated message. If you feel this action was in error, please comment on this issue so it can be looked at again*

**Comment by Dravere** [comment]
@julianlam I highly disagree with your bot on this :wink:

**Comment by nodebb-misty** [comment]
Issue closed due to inactivity.

----

*This is an automated message. If you feel this action was in error, please comment on this issue so it can be looked at again*

**Comment by Dravere** [comment]
@julianlam Why exactly is misty closing this issue? Because of the "needs confirmation" tag? Do we actually really still need confirmation? I mean we are able to reproduce it, we just don't know exactly what is causing it.

Also it might be an idea to further forward the AJAX error and also check the `readyState` of the `XMLHttpRequest`. If the `readyState` is 0 and `status` is 0 there likely is a network error and the `status` shouldn't be changed to 500.

**Comment by pitaj** [comment]
It shouldn't, I doubt there is any way to reproduce that.

**Comment by Dravere** [comment]
I just spent some time too look into it further. And this bug isn't getting less weird.

I took a look at an instance that was displaying the loading icon and no internal error appeared in Vivaldi (V8 6.4.388.46, Chrome/64.0.3282.189). I saw that there is an error before the script error posted above:
```
Unable to load template: 500
(anonymous) @ ajaxify.js:333
(anonymous) @ require.js:7
(anonymous) @ require.js:23
y @ require.js:6
emit @ require.js:23
(anonymous) @ require.js:11
y @ require.js:6
l @ require.js:11
onScriptError @ require.js:30
```
This probably isn't helping much. But it seems to be closely related to RequireJS. But at the end it always comes from a script error in the 500 template. Which is really weird.

But I noticed I haven't tested it with a different browser yet. I tested the instance that showed immediately the "Internal Error" with a different browser, but not the ones displaying the loading icon.

So I loaded up the latest Firefox (59.0.2) and tested on the instance with the loading icon. Sure enough it works perfectly. It shows the loading icon and then a message appears informing me that there seems to be no connection to the forum. This message doesn't appear at all in Vivaldi. But after I re-enabled the network connection again, I received the "Internal Error" message. So basically Firefox wasn't able to load the 500 template, displayed that there was no connection and as soon as there was a connection again it displayed the 500 error.

But when I went into the network log of Firefox, I noticed that it seemed to have loaded a cached 500 template. When I disabled the cache and redid the entire thing, it worked perfectly and displayed not internal error. My guess is that it never defined the 500 template from the main html page.

---

As confusing this entire thing is, I feel like there are two separate issues here:

1. NodeBB tries to display the 500 template in case of no connectivity. Which isn't helpful and should not be the case. Which is this issue originally.
2. NodeBB in some cases has no local definition of the 500 template even though it should have been defined in the main html. Which is a new issue that appeared.

I think if we fix the first issue the second issue might become less important. Since even if it wouldn't load the 500 template via the definition in the main html, if it has network connection and this template isn't used for the no network connection error, it could load it via the network.

**Comment by Dravere** [comment]
After writing the last post it hit me. I think I know what is going on. Or at least I think I know why they can't find the definition. Whenever I've had an instance that displayed the loading icon it didn't reside at the root path.

The defintion in the main html is: `define('/assets/templates/500.js', ...`.  
It doesn't respect the relative path. But the load template of ajaxify does respect it:
```
ajaxify.loadTemplate = function (template, callback) {
    require([config.relative_path + '/assets/templates/' + template + '.js'], callback, function (err) {
        console.error('Unable to load template: ' + template);
        throw err;
    });
};
```

@julianlam Could it be that your local dev environment also isn't at root level?

This would explain the second issue. But fixing it will cause that every NodeBB will display the internal error message when there is no connection. So I would suggest fixing the reported issue (first one) at the same time.

**Comment by julianlam** [comment]
@Dravere Yes, that's correct. I'm the only one of the main devs that regularly develops on a subfolder. That is likely a significant finding... :think

**Comment by barisusakli** [comment]
@Dravere did the above commit fix your issue?

**Comment by Dravere** [comment]
It fixed the inconsistent behavior. But now it will always show an internal error when losing connection. Like I said previously, it fixed the second issue but not the first one that was initially reported in this issue.

So with this fix you should now be able to reproduce the error on your side. Since as far as I understand it, your development is at the root level?
